File: plugins/cicero/hooks/language-nudge.py
#!/usr/bin/env python3
"""UserPromptSubmit hook — re-assert the user's CURRENT language every turn.

SessionStart injects the voice once per session; over a long session the model can drift,
and the user may switch languages mid-session (SessionStart can't react to that). This
fires on each prompt, detects the prompt's dominant script (latin / cyrillic / cjk), and
injects a one-line reminder to reply in that language (CICERO 12). Cheap (~15 tokens), and
silent when the script is Latin or ambiguous — English is the model default and needs no
nudge. The static voice rules ship as an output style; this hook only keeps the reply language current.
"""
import json
import sys

SCRIPT_LABEL = {
    "cyrillic": "Cyrillic (e.g. Russian)",
    "cjk": "CJK (Chinese / Japanese / Korean)",
}


def dominant_script(s):
    """Dominant script family of the text, or None if none clearly leads."""
    lat = cyr = cjk = 0
    for ch in s:
        o = ord(ch)
        if "a" <= ch.lower() <= "z":
            lat += 1
        elif 0x0400 <= o <= 0x04FF:
            cyr += 1
        elif (0x3040 <= o <= 0x30FF) or (0x3400 <= o <= 0x9FFF) or (0xAC00 <= o <= 0xD7AF):
            cjk += 1
    counts = {"latin": lat, "cyrillic": cyr, "cjk": cjk}
    name = max(counts, key=counts.get)
    if counts[name] >= 3 and counts[name] > sum(v for k, v in counts.items() if k != name):
        return name
    return None


def main():
    try:
        data = json.load(sys.stdin)
    except Exception:
        sys.exit(0)
    prompt = data.get("prompt", "")
    if not isinstance(prompt, str) or not prompt.strip():
        sys.exit(0)

    script = dominant_script(prompt)
    if script in SCRIPT_LABEL:
        ctx = (
            f"[CICERO 12] The user's latest message is in {SCRIPT_LABEL[script]} script — "
            "reply in that language. Keep code, paths, and identifiers in English."
        )
        print(json.dumps(
            {"hookSpecificOutput": {"hookEventName": "UserPromptSubmit", "additionalContext": ctx}},
            ensure_ascii=False,
        ))
    sys.exit(0)


if __name__ == "__main__":
    main()

File: plugins/cicero/hooks/language-nudge.js
#!/usr/bin/env node
// UserPromptSubmit hook — re-assert the user's CURRENT language every turn (CICERO 7).
//
// SessionStart injects the voice once per session; over a long session the model can
// drift, and the user may switch languages mid-session. This fires on each prompt,
// detects the prompt's dominant script (latin / cyrillic / cjk), and injects a one-line
// reminder to reply in that language. Cheap (~15 tokens), silent when the script is
// Latin or ambiguous — English is the model default and needs no nudge.
//
// Node rewrite of the former language-nudge.py: unicode codepoint classification is
// clean and portable in node (~10ms startup) but fragile in bash on macOS, whose grep
// lacks \p{Script}. Hooks are shell-first here; this one is the deliberate node
// exception for unicode — rationale in docs/adr/0002. Fail-open: any error exits 0.

const LABEL = {
  cyrillic: 'Cyrillic (e.g. Russian)',
  cjk: 'CJK (Chinese / Japanese / Korean)',
};

function dominantScript(text) {
  let latin = 0;
  let cyrillic = 0;
  let cjk = 0;
  for (const ch of text) {
    const code = ch.codePointAt(0);
    const lower = ch.toLowerCase();
    if (lower >= 'a' && lower <= 'z') {
      latin += 1;
    } else if (code >= 0x0400 && code <= 0x04ff) {
      cyrillic += 1;
    } else if (
      (code >= 0x3040 && code <= 0x30ff) ||
      (code >= 0x3400 && code <= 0x9fff) ||
      (code >= 0xac00 && code <= 0xd7af)
    ) {
      cjk += 1;
    }
  }
  const counts = { latin, cyrillic, cjk };
  // First key wins ties (latin > cyrillic > cjk), matching python's max(counts, key=...).
  let name = 'latin';
  for (const key of Object.keys(counts)) {
    if (counts[key] > counts[name]) {
      name = key;
    }
  }
  const others = Object.keys(counts).reduce(
    (sum, key) => (key === name ? sum : sum + counts[key]),
    0,
  );
  if (counts[name] >= 3 && counts[name] > others) {
    return name;
  }
  return null;
}

let raw = '';
process.stdin.setEncoding('utf8');
process.stdin.on('data', (chunk) => {
  raw += chunk;
});
process.stdin.on('end', () => {
  try {
    const prompt = JSON.parse(raw).prompt || '';
    if (typeof prompt === 'string' && prompt.trim()) {
      const script = dominantScript(prompt);
      if (script && LABEL[script]) {
        const context =
          `[CICERO 7] The user's latest message is in ${LABEL[script]} script — ` +
          'reply in that language. Keep code, paths, and identifiers in English.';
        process.stdout.write(
          `${JSON.stringify({
            hookSpecificOutput: {
              hookEventName: 'UserPromptSubmit',
              additionalContext: context,
            },
          })}\n`,
        );
      }
    }
  } catch (error) {
    // fail-open — never block a prompt over a nudge
  }
  process.exit(0);
});
